oak_sign/src/main.rs

```diff
@@ -0,0 +1,197 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! An utility binary to sign Oak Wasm modules.
+//!
+//! Invoke with:
+//!
+//! ```shell
+//! cargo run --manifest-path=oak_sign/Cargo.toml -- \
+//!     generate \
+//!     --private-key=<PRIVATE_KEY_FILE>.key \
+//!     --public-key=<PUBLIC_KEY_FILE>.pem
+//!
+//! cargo run --manifest-path=oak_sign/Cargo.toml -- \
+//!     sign \
+//!     --private-key=<PRIVATE_KEY_FILE>.key \
+//!     --module=<WASM_MODULE_FILE>.wasm \
+//!     --signature=<SIGNATURE_FILE>.sign
+//!
+//! cargo run --manifest-path=oak_sign/Cargo.toml -- \
+//!     verify \
+//!     --module=<WASM_MODULE_FILE>.wasm \
+//!     --signature=<SIGNATURE_FILE>.sign
+//! ```
+
+use anyhow::Context;
+use log::info;
+use pem::{encode, encode_many, parse, parse_many, Pem};
+use ring::{
+    rand,
+    signature::{self, KeyPair},
+};
+use std::{
+    collections::HashMap,
+    fs::{read, write},
+};
+use structopt::StructOpt;
+
+/// Command line options for the Oak sign.
+#[derive(StructOpt, Clone)]
+#[structopt(about = "Oak Sign")]
+struct Opt {
+    #[structopt(subcommand)]
+    cmd: Command,
+}
+
+/// Available commands for the Oak sign.
+#[derive(StructOpt, Clone)]
+enum Command {
+    Generate(Generate),
+    Sign(Sign),
+    Verify(Verify),
+}
+
+#[derive(StructOpt, Clone)]
+#[structopt(about = "Generate Ed25519 key pair")]
+struct Generate {
+    #[structopt(long, help = "Output private key file path")]
+    pub private_key: String,
+    #[structopt(long, help = "Output public key file path")]
+    pub public_key: String,
+}
+
+#[derive(StructOpt, Clone)]
+#[structopt(about = "Sign Wasm module")]
+struct Sign {
+    #[structopt(long, help = "Input private key file path")]
+    pub private_key: String,
+    #[structopt(long, help = "Input Wasm module path")]
+    pub module: String,
```

Comment:
Is this specific about Wasm? AFAICT it just signs an arbitrary file, so I think the naming should be more generic. For instance, it may be used to sign other things in the future if we wanted to. So maybe just `--input` or `--input_file`?

Reply:
Done